core/src/main/java/org/apache/accumulo/fate/Fate.java

```diff
@@ -71,24 +83,43 @@ public void run() {
         try {
           tid = store.reserve();
           TStatus status = store.getStatus(tid);
+          if (status == FAILED) {
+            runnerLog.info(
+                "FATE txid " + Long.toHexString(tid) + " likely cancelled before it started.");
+            return;
+          }
           Repo<T> op = store.top(tid);
-          if (status == TStatus.FAILED_IN_PROGRESS) {
+          if (status == FAILED_IN_PROGRESS) {
             processFailed(tid, op);
           } else {
             Repo<T> prevOp = null;
             try {
-              deferTime = op.isReady(tid, environment);
-
-              if (deferTime == 0) {
-                prevOp = op;
-                if (status == TStatus.SUBMITTED) {
-                  store.setStatus(tid, TStatus.IN_PROGRESS);
+              try {
+                synchronized (RUNNING_TRANSACTIONS) {
+                  RUNNING_TRANSACTIONS.put(tid, Thread.currentThread());
                 }
-                op = op.call(tid, environment);
-              } else
-                continue;
-
+                deferTime = op.isReady(tid, environment);
+                if (deferTime == 0) {
+                  prevOp = op;
+                  if (status == SUBMITTED) {
+                    store.setStatus(tid, IN_PROGRESS);
+                  }
+                  op = op.call(tid, environment);
+                } else
+                  continue;
+              } finally {
+                synchronized (Thread.currentThread()) {
+                  synchronized (RUNNING_TRANSACTIONS) {
+                    RUNNING_TRANSACTIONS.remove(tid);
+                  }
+                }
```

Comment:
```suggestion
                  synchronized (RUNNING_TRANSACTIONS) {
                    RUNNING_TRANSACTIONS.remove(tid);
                  }
                  // TODO maybe we should check/clear for interrupt here... if something 
                  // was trying to interrupt right before the sync block then we do not want 
                  // that to impact the next thing this thread will work on....
```

Reply:
Incorporated this change into [c49f55e](https://github.com/apache/accumulo/pull/2467/commits/c49f55ec90ed3ea883c9dbfd22dfac740b3e8b3d)